C error. mut type cast of unwrapped interface

### Describe the bug

`src/message/message.v`

```v
module message

pub struct Message {
pub mut:
	text string = 'Init'
	m2   ?&Message
}

pub fn (mut m Message) update_text(new_text string) {
	m.text = new_text
}

pub fn (mut m Message) update_ext_text(new_text string) {
	m.m2?.text = new_text
}
```

`src/main.v`

```v
module main

import time
import message

// Build this example with `v -live .`

type LocalMessageType = message.Message

interface IMessage {
mut:
	text string
	update_text(new_text string)
}

struct App {
mut:
	message &IMessage
}

pub fn (mut m LocalMessageType) update_text_externally(new_text string) {
	m.text = new_text
}

@[live]
fn print_message(mut mdef IMessage) {
	if mut mdef is message.Message {
		// mut m := LocalMessageType(*mdef) // works. with -live: Unhandled Exception 0xC0000005
		// mut m := &LocalMessageType(mdef) // works, warning. with -live: Unhandled Exception 0xC0000005
		mut m := LocalMessageType(mdef) // does not work, C compilation error

		dump(mdef)
		dump(*mdef)
		dump(m)

		m.update_text('Updated within')
		m.update_ext_text('Updated ext')
		m.update_text_externally('Updated externally')
		println('${m.text}, ${m.m2?.text}')
	}
}

fn main() {
	for {
		mut m2 := &message.Message{
			text: 'Init External'
		}
		mut mdef := &message.Message{
			text: 'Init'
			m2:   m2
		}
		mut app := &App{
			message: mdef
		}
		print_message(mut app.message)
		time.sleep(500 * time.millisecond)
	}
}
```

```
> v run .      
INFO: compile with `v -live . `, if you want to use the @[live] function main.print_message .
================== C compilation error (from tcc): ==============
cc: C:/Users/phcre/AppData/Local/Temp/v_0/vtests.01JDYPQX81M8CS5V5Q38G0PNJC.tmp.c:7803: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/phcre/AppData/Local/Temp/v_0/vtests.01JDYPQX81M8CS5V5Q38G0PNJC.tmp.c:17195: error: cannot convert 'struct src__message__Message *' to 'struct src__message__Message'
... (the original output was 3 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Reproduction Steps

Type cast an unwrapped interface

### Expected Behavior

A checker warning

### Current Behavior

C compile error

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.8 5092626.e4d24e8

### Environment details (OS name and version, etc.)

```
> v doctor
V full version: V 0.4.8 5092626.e4d24e8
OS: windows, Microsoft Windows 10 Pro v19045 64-bit
Processor: 8 cpus, 64bit, little endian, 

getwd: C:\Users\phcre\Documents\v\vtests
vexe: C:\Users\phcre\Documents\v\v\v.exe
vexe mtime: 2024-11-24 01:17:49

vroot: OK, value: C:\Users\phcre\Documents\v\v
VMODULES: OK, value: C:\Users\phcre\.vmodules
VTMP: OK, value: C:\Users\phcre\AppData\Local\Temp\v_0

Git version: git version 2.24.1.windows.2
Git vroot status: weekly.2024.37-553-ge4d24e89 (40 commit(s) behind V master)
.git/config present: true

CC version: cc (Rev6, Built by MSYS2 project) 13.1.0
emcc version: emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.44 (bec42dac7873903d09d713963e34020c22a8bd2d)
thirdparty/tcc status: thirdparty-windows-amd64 b425ac82
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzRiMWU0MmQyZWY0Y2JjYzQ2Y2JhZTEiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.WpOf46seKXonqq4nbtg95Uq6J6QUdhkGxPbDld05YkI">Huly&reg;: <b>V_0.6-21458</b></a></sub>

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3614,9 +3614,8 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		c.error('cannot cast `${ft}` ${kind_name} value to `${tt}`, use `${node.expr} as ${tt}` instead',
 			node.pos)
 	}
-	if from_sym.language == .v && from_type.is_ptr() && !to_type.is_ptr() && !to_type.is_int()
-		&& to_type != ast.voidptr_type && !node.expr.is_auto_deref_var()
-		&& final_to_sym.kind != .interface {
+	if from_sym.language == .v && from_type.is_ptr() && !to_type.is_ptr()
+		&& !node.expr.is_auto_deref_var() && final_to_sym.kind == .struct {
 		ft := c.table.type_to_str(from_type)
 		tt := c.table.type_to_str(to_type)
 		c.error('cannot cast `${ft}` to `${tt}`, you must dereference it first (e.g. ${tt}(*var))',

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3614,6 +3614,14 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		c.error('cannot cast `${ft}` ${kind_name} value to `${tt}`, use `${node.expr} as ${tt}` instead',
 			node.pos)
 	}
+	if from_sym.language == .v && from_type.is_ptr() && !to_type.is_ptr() && !to_type.is_int()
+		&& to_type != ast.voidptr_type && !node.expr.is_auto_deref_var()
+		&& final_to_sym.kind != .interface {
+		ft := c.table.type_to_str(from_type)
+		tt := c.table.type_to_str(to_type)
+		c.error('cannot cast `${ft}` to `${tt}`, you must dereference it first (e.g. ${tt}(*var))',
+			node.pos)
+	}
 
 	if node.has_arg {
 		c.expr(mut node.arg)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3614,7 +3614,7 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		c.error('cannot cast `${ft}` ${kind_name} value to `${tt}`, use `${node.expr} as ${tt}` instead',
 			node.pos)
 	}
-	if from_sym.language == .v && from_type.is_ptr() && !to_type.is_ptr()
+	if from_sym.language == .v && from_type.is_ptr() && !final_to_type.is_ptr()
 		&& !node.expr.is_auto_deref_var() && final_to_sym.kind == .struct
 		&& final_from_sym.kind == .struct {
 		if c.check_struct_signature(final_from_sym.info as ast.Struct, final_to_sym.info as ast.Struct) {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3615,11 +3615,14 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 			node.pos)
 	}
 	if from_sym.language == .v && from_type.is_ptr() && !to_type.is_ptr()
-		&& !node.expr.is_auto_deref_var() && final_to_sym.kind == .struct {
-		ft := c.table.type_to_str(from_type)
-		tt := c.table.type_to_str(to_type)
-		c.error('cannot cast `${ft}` to `${tt}`, you must dereference it first (e.g. ${tt}(*var))',
-			node.pos)
+		&& !node.expr.is_auto_deref_var() && final_to_sym.kind == .struct
+		&& final_from_sym.kind == .struct {
+		if c.check_struct_signature(final_from_sym.info as ast.Struct, final_to_sym.info as ast.Struct) {
+			ft := c.table.type_to_str(from_type)
+			tt := c.table.type_to_str(to_type)
+			c.error('cannot cast `${ft}` to `${tt}`, you must dereference it first (e.g. ${tt}(*var))',
+				node.pos)
+		}
 	}
 
 	if node.has_arg {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3614,7 +3614,7 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 		c.error('cannot cast `${ft}` ${kind_name} value to `${tt}`, use `${node.expr} as ${tt}` instead',
 			node.pos)
 	}
-	if from_sym.language == .v && from_type.is_ptr() && !final_to_type.is_ptr()
+	if from_sym.language == .v && from_type.is_ptr() && !to_type.is_ptr() && !final_to_type.is_ptr()
 		&& !node.expr.is_auto_deref_var() && final_to_sym.kind == .struct
 		&& final_from_sym.kind == .struct {
 		if c.check_struct_signature(final_from_sym.info as ast.Struct, final_to_sym.info as ast.Struct) {

--- a/vlib/v/checker/tests/cast_to_concrete_mut_err.out
+++ b/vlib/v/checker/tests/cast_to_concrete_mut_err.out
@@ -1,13 +1,13 @@
 vlib/v/checker/tests/cast_to_concrete_mut_err.vv:37:12: warning: casting `&Message` to `&LocalMessageType` is only allowed in `unsafe` code
    35 |     if mut mdef is Message {
-   36 |         mut a := LocalMessageType(*mdef) // works. with -live: Unhandled Exception 0xC0000005
-   37 |         mut b := &LocalMessageType(mdef) // works, warning. with -live: Unhandled Exception 0xC0000005
+   36 |         mut a := LocalMessageType(*mdef)
+   37 |         mut b := &LocalMessageType(mdef)
       |                  ~~~~~~~~~~~~~~~~~~~~~~~
    38 |         mut c := LocalMessageType(mdef)
    39 |
 vlib/v/checker/tests/cast_to_concrete_mut_err.vv:38:12: error: cannot cast `&Message` to `LocalMessageType`, you must dereference it first (e.g. LocalMessageType(*var))
-   36 |         mut a := LocalMessageType(*mdef) // works. with -live: Unhandled Exception 0xC0000005
-   37 |         mut b := &LocalMessageType(mdef) // works, warning. with -live: Unhandled Exception 0xC0000005
+   36 |         mut a := LocalMessageType(*mdef)
+   37 |         mut b := &LocalMessageType(mdef)
    38 |         mut c := LocalMessageType(mdef)
       |                  ~~~~~~~~~~~~~~~~~~~~~~
    39 | 

--- a/vlib/v/checker/tests/cast_to_concrete_mut_err.out
+++ b/vlib/v/checker/tests/cast_to_concrete_mut_err.out
@@ -1,0 +1,14 @@
+vlib/v/checker/tests/cast_to_concrete_mut_err.vv:37:12: warning: casting `&Message` to `&LocalMessageType` is only allowed in `unsafe` code
+   35 |     if mut mdef is Message {
+   36 |         mut a := LocalMessageType(*mdef) // works. with -live: Unhandled Exception 0xC0000005
+   37 |         mut b := &LocalMessageType(mdef) // works, warning. with -live: Unhandled Exception 0xC0000005
+      |                  ~~~~~~~~~~~~~~~~~~~~~~~
+   38 |         mut c := LocalMessageType(mdef)
+   39 |
+vlib/v/checker/tests/cast_to_concrete_mut_err.vv:38:12: error: cannot cast `&Message` to `LocalMessageType`, you must dereference it first (e.g. LocalMessageType(*var))
+   36 |         mut a := LocalMessageType(*mdef) // works. with -live: Unhandled Exception 0xC0000005
+   37 |         mut b := &LocalMessageType(mdef) // works, warning. with -live: Unhandled Exception 0xC0000005
+   38 |         mut c := LocalMessageType(mdef)
+      |                  ~~~~~~~~~~~~~~~~~~~~~~
+   39 | 
+   40 |         dump(a)

--- a/vlib/v/checker/tests/cast_to_concrete_mut_err.vv
+++ b/vlib/v/checker/tests/cast_to_concrete_mut_err.vv
@@ -1,0 +1,58 @@
+module main
+
+pub struct Message {
+pub mut:
+	text string = 'Init'
+	m2   ?&Message
+}
+
+pub fn (mut m Message) update_text(new_text string) {
+	m.text = new_text
+}
+
+pub fn (mut m Message) update_ext_text(new_text string) {
+	m.m2?.text = new_text
+}
+
+type LocalMessageType = Message
+
+interface IMessage {
+mut:
+	text string
+	update_text(new_text string)
+}
+
+struct App {
+mut:
+	message &IMessage
+}
+
+pub fn (mut m LocalMessageType) update_text_externally(new_text string) {
+	m.text = new_text
+}
+
+fn print_message(mut mdef IMessage) {
+	if mut mdef is Message {
+		mut a := LocalMessageType(*mdef) // works. with -live: Unhandled Exception 0xC0000005
+		mut b := &LocalMessageType(mdef) // works, warning. with -live: Unhandled Exception 0xC0000005
+		mut c := LocalMessageType(mdef)
+
+		dump(a)
+		dump(b)
+		dump(c)
+	}
+}
+
+fn test_main() {
+	mut m2 := &Message{
+		text: 'Init External'
+	}
+	mut mdef := &Message{
+		text: 'Init'
+		m2:   m2
+	}
+	mut app := &App{
+		message: mdef
+	}
+	print_message(mut app.message)
+}

--- a/vlib/v/checker/tests/cast_to_concrete_mut_err.vv
+++ b/vlib/v/checker/tests/cast_to_concrete_mut_err.vv
@@ -33,8 +33,8 @@ pub fn (mut m LocalMessageType) update_text_externally(new_text string) {
 
 fn print_message(mut mdef IMessage) {
 	if mut mdef is Message {
-		mut a := LocalMessageType(*mdef) // works. with -live: Unhandled Exception 0xC0000005
-		mut b := &LocalMessageType(mdef) // works, warning. with -live: Unhandled Exception 0xC0000005
+		mut a := LocalMessageType(*mdef)
+		mut b := &LocalMessageType(mdef)
 		mut c := LocalMessageType(mdef)
 
 		dump(a)

